[WASM GC] Fix BBQJIT::addRefI31 on constant expression

dhecht opened the pull request:
#### 06e92c10a5725ff226b8ba6a00cc428614d781a0
<pre>
[WASM GC] Fix BBQJIT::addRefI31 on constant expression
<a href="https://bugs.webkit.org/show_bug.cgi?id=293207">https://bugs.webkit.org/show_bug.cgi?id=293207</a>
<a href="https://rdar.apple.com/151653787">rdar://151653787</a>

Reviewed by Yusuke Suzuki.

The constant evaluation path is constructing the final
i32.ref value incorrectly. The lhs of | is promoted to
signed 64-bits. So, when the constant had bit 30 set
(high bit of the i31), the sign was extended all the way
to 64 bits, rather than 32 bits.

This would not be noticible when doing i31.get_{u, s}
but is noticible when using the value as a ref, e.g. ref.eq.

Bonus: cleanup in InPlaceInterpreter64.asm is a no-op
but the value being pushed is the 64-bit ref, so use pushQuad().

Canonical link: <a href="https://commits.webkit.org/295271@main">https://commits.webkit.org/295271@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b57edb168ddc7da2633c6b3f0de3ec735cb5e96

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104545 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24255 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14673 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109754 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55216 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24640 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32799 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79371 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107551 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19160 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94353 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59697 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18941 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12425 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54587 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/97224 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88641 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12480 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112142 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/103160 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31706 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23342 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88469 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32070 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90584 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88085 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22449 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32982 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10755 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/27023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31633 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36973 "Failed to build and analyze WebKit") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/127441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31425 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/127441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34764 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32985 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
```diff
@@ -1502,7 +1502,8 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::truncSaturated(Ext1OpType truncationOp,
-        result = Value::fromI64((((value.asI32() & 0x7fffffff) << 1) >> 1) | JSValue::NumberTag);
+        uint32_t lo32 = (value.asI32() << 1) >> 1;
+        result = Value::fromI64(lo32 | JSValue::NumberTag);
         LOG_INSTRUCTION("RefI31", value, RESULT(result));
```
I wonder if we should update the JIT side code too. Also can you check what is done in OMG?

dhecht replied:
OMG is same encoding as the non-constant BBQ and interpreters. It was just this constant folding that disagreed.

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -3619,7 +3619,7 @@ ipintOp(_ref_i31, macro()
     lshifti 0x1, t0
     rshifti 0x1, t0
     orq TagNumber, t0
```
This code should be updated to be aligned to BBQJIT code.

dhecht replied:
This is equivalent, no? I had checked the interpreters and OMG, and the only thing that didn't match was the BBQ constant case, so I changed that to be the same as these. (They all have a superfluous `and 0x7fffffff` but I was going to clean that up separately since it's not incorrect.)

The original test case compared IPInt with BBQ by using a v128 in one of the functions that constructs the i31ref (forcing it into that broken BBQ path), but I didn't want to rely on that since it will stop verifying once IPInt supports SIMD. But that does show that this produces the right encoding.

dhecht committed:
[WASM GC] Fix BBQJIT::addRefI31 on constant expression
https://bugs.webkit.org/show_bug.cgi?id=293207
rdar://151653787

Reviewed by Yusuke Suzuki.

The constant evaluation path is constructing the final
i32.ref value incorrectly. The lhs of | is promoted to
signed 64-bits. So, when the constant had bit 30 set
(high bit of the i31), the sign was extended all the way
to 64 bits, rather than 32 bits.

This would not be noticible when doing i31.get_{u, s}
but is noticible when using the value as a ref, e.g. ref.eq.

Bonus: cleanup in InPlaceInterpreter64.asm is a no-op
but the value being pushed is the 64-bit ref, so use pushQuad().

Canonical link: https://commits.webkit.org/295271@main
